client/data/emails/use-add-email-forward-mutation.ts

```diff
@@ -0,0 +1,194 @@
+
+		await queryClient.cancelQueries( useGetEmailAccountsQueryKey );
+		await queryClient.cancelQueries( useGetDomainsQueryKey );
+
+		const previousEmailAccountsQueryData = queryClient.getQueryData< any >(
+			useGetEmailAccountsQueryKey
+		);
+		const emailForwards = previousEmailAccountsQueryData?.accounts?.[ 0 ]?.emails;
+
+		// Optimistically add email forward to `useGetEmailAccountsQuery` data
+		if ( emailForwards ) {
+			const newEmailForwards = orderBy(
+				[
+					...emailForwards,
+					{
+						domain: domainName,
+						email_type: 'email_forward',
+						is_verified: false,
+						mailbox,
+						role: 'standard',
+						target: destination,
+						temporary: true,
+						warnings: [],
+					},
+				],
+				[ 'mailbox' ],
+				[ 'asc' ]
+			);
+
+			queryClient.setQueryData( useGetEmailAccountsQueryKey, {
+				...previousEmailAccountsQueryData,
+				accounts: [
+					{
+						...previousEmailAccountsQueryData.accounts[ 0 ],
+						emails: newEmailForwards,
+					},
+				],
+			} );
+		}
+
+		const previousDomainsQueryData = queryClient.getQueryData< any >( useGetDomainsQueryKey );
+
+		// Optimistically increment the email_forward counter in `useGetDomainsQuery` data
+		if ( previousDomainsQueryData ) {
+			const selectedDomainIndex = previousDomainsQueryData.domains.findIndex(
+				( { domain }: { domain: string } ) => domain === domainName
+			);
+
+			if ( selectedDomainIndex >= 0 ) {
+				const selectedDomain = previousDomainsQueryData.domains[ selectedDomainIndex ];
+				const newDomains = [ ...previousDomainsQueryData.domains ];
+
+				newDomains.splice( selectedDomainIndex, 1, {
+					...selectedDomain,
+					email_forwards_count: selectedDomain.email_forwards_count + 1,
+				} );
+
+				queryClient.setQueryData( useGetDomainsQueryKey, {
+					...previousDomainsQueryData,
+					domains: newDomains,
+				} );
+			}
+		}
+
+		return {
+			[ JSON.stringify( useGetEmailAccountsQueryKey ) ]: previousEmailAccountsQueryData,
+			[ JSON.stringify( useGetDomainsQueryKey ) ]: previousDomainsQueryData,
+		};
+	};
+
+	mutationOptions.onError = ( error, variables, context ) => {
+		suppliedOnError?.( error, variables, context );
+
+		if ( context ) {
+			queryClient.setQueryData(
+				useGetEmailAccountsQueryKey,
+				context[ JSON.stringify( useGetEmailAccountsQueryKey ) ]
+			);
+
+			queryClient.setQueryData(
+				useGetDomainsQueryKey,
+				context[ JSON.stringify( useGetDomainsQueryKey ) ]
+			);
+		}
+
+		let errorMessage = translate(
+			'Failed to add email forwarding record. ' +
+				'Please try again or ' +
+				'{{contactSupportLink}}contact support{{/contactSupportLink}}.',
+			{
+				components: {
+					contactSupportLink: createElement( 'a', { href: CALYPSO_CONTACT } ),
+				},
+			}
+		);
+
+		if ( error ) {
+			errorMessage = translate(
+				'Failed to add email forwarding record ' +
+					'with message "%(message)s". ' +
+					'Please try again or ' +
+					'{{contactSupportLink}}contact support{{/contactSupportLink}}.',
```

Comment:
I think we could take advantage of this pull request to refine those messages a bit:

```suggestion
				'Failed to add email forwards ' +
					'with message "%(message)s". ' +
					'Please try again or ' +
					'{{contactSupportLink}}contact support{{/contactSupportLink}}.',
```

That way they are more consistent with the `Add New Email Forwards` heading shown on the page:

![image](https://user-images.githubusercontent.com/594356/174128184-f6b7ca4f-563d-41f0-9c03-ac5353462fe8.png)

Reply:
Good point! Fixed it.